### PR TITLE
Create proposal

### PR DESCRIPTION
**Pull request by tbwebb22**

### CODE REVIEW

**Comment by adamgall** [comment]
@christopherdancy when you have time today, can you please have the first go at doing code review on this PR?

just come at it from the point of view of "what would Adam's stickler-ass say"

**Comment by adamgall** [comment]
@tbwebb22 please link this PR with the appropriate Issue

File: src/components/DAOs/DAO/CreateProposal/Essentials.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  proposalDescription,
+  setProposalDescription,
+}: {
+  proposalDescription: string;
+  setProposalDescription: React.Dispatch<React.SetStateAction<string>>;
+}) => {
+  return (
+    <div className="mx-auto bg-slate-100 px-8 mt-4 mb-8 pt-8 pb-8 content-center">
+      <div className="pb-8 text-lg">Essentials</div>
+      <CreateDAOInput
```
Comment (christopherdancy) [inline src/components/DAOs/DAO/CreateProposal/Essentials.tsx:14]: More General Component Naming

Comment (christopherdancy) [reply src/components/DAOs/DAO/CreateProposal/Essentials.tsx:14]: i.e- multistepInput

File: src/components/DAOs/DAO/CreateProposal/Transaction.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      targetAddress: transaction.targetAddress,
+      functionName: transaction.functionName,
+      functionSignature: functionSignature,
+      parameters: transaction.parameters,
+    };
+
+    updateTransaction(newTransactionData, transactionNumber);
+  };
+
+  const updateParameters = (parameters: string) => {
```
Comment (christopherdancy) [inline src/components/DAOs/DAO/CreateProposal/Transaction.tsx:54]: This makes sense. If there are multiple transactions. a user can update anyone of them.

File: src/components/DAOs/DAO/CreateProposal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import { useEffect, useState } from "react";
+import Button from "../../../ui/Button";
+import { BigNumber, ethers } from "ethers";
+import useCreateProposal from "../../../../daoData/useCreateProposal";
+import ConnectModal from '../../../ConnectModal';
+import Pending from '../../../Pending';
+import Essentials from './Essentials';
+import Transactions from './Transactions';
+
+export type TransactionData = {
```
Comment (christopherdancy) [inline src/components/DAOs/DAO/CreateProposal/index.tsx:10]: Could this be a type on the transactions component which is pulled in to this component for types?

File: src/components/DAOs/DAO/CreateProposal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    },
+  ]);
+  const [pending, setPending] = useState<boolean>(false);
+  const [proposalData, setProposalData] = useState<ProposalData>();
+
+  const addTransaction = () => {
+    setTransactions([
+      ...transactions,
+      {
+        targetAddress: "",
```
Comment (christopherdancy) [inline src/components/DAOs/DAO/CreateProposal/index.tsx:42]: There should be state checking for each input.  Check below for an example found in DAODetails.tsx

`  useEffect(() => {
    setNextEnabled(name !== undefined && name.trim() !== "")
  }, [name, setNextEnabled]);`

File: src/components/DAOs/DAO/CreateProposal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+  return (
+    <div>
+      <Pending message="Creating Proposal..." pending={pending} />
+      <ConnectModal />
+      <div className="mx-24">
+        <div className="mt-24 text-xl">Create Proposal</div>
+
+        <div className="flex">
+          {step === 0 && (
```
Comment (christopherdancy) [inline src/components/DAOs/DAO/CreateProposal/index.tsx:97]: Review stepDisplay component on CreateDAO/index.tsx. Creating a component would be cleaner

**Review by christopherdancy — COMMENTED** [review]
Please review @adamgall

File: src/components/DAOs/DAO/index.tsx [hunk 6]
```diff
@@ -14,6 +15,7 @@ function DAOs() {
     <Routes>
       <Route index element={<Summary />} />
       <Route path="details" element={<Details address={params.address} />} />
+      <Route path="createproposal" element={<CreateProposal address={params.address} />} />
```
Comment (adamgall) [inline src/components/DAOs/DAO/index.tsx:18]: @tbwebb22

to keep these routes RESTful (and give some "breathing room" for the immediate next work to come with the "proposals" work), the route for this should be `/daos/:address/proposals/new`

So, can you create a directory in the tree for `Proposals`, with an `index.tsx` that defines the routes (with just `new` for now), and a directory in there called `New` containing evrything that's currently in the `CreateProposal` directory? Basically, follow the pattern that the DAO directory has started.

Thanks!

File: src/components/DAOs/DAO/Summary.tsx [hunk 7]
```diff
@@ -29,7 +29,12 @@ function ValidDAO({
         is a valid dao!
       </H1>
       <div>
-        <Link to="details">DAO Details</Link>
+        <div>
+        <Link to="details" className="underline">DAO Details</Link>
```
Comment (adamgall) [inline src/components/DAOs/DAO/Summary.tsx:33]: nitpick: please make sure things are indented properly in here

File: src/components/DAOs/DAO/CreateProposal/Transaction.tsx [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    const newTransactionData = {
+      targetAddress: transaction.targetAddress,
+      functionName: transaction.functionName,
+      functionSignature: transaction.functionSignature,
+      parameters: parameters,
+    };
+
+    updateTransaction(newTransactionData, transactionNumber);
+  };
```
Comment (adamgall) [inline src/components/DAOs/DAO/CreateProposal/Transaction.tsx:63]: Wondering if there's a way to DRY up these functions. Not asking anything of you here, just typing out loud.

**Review by adamgall — CHANGES_REQUESTED** [review]
Just some small changes. Biggest thing is related to RESTful routing

**Comment by tbwebb22** [comment]
@adamgall Addressed these comments in recent commits

File: src/components/DAOs/DAO/Proposals/index.tsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+function Proposals({ address }: { address: string }) {
+  const params = useParams();
+
+  if (!params.address) {
+    return <div>if you see this, it's a bug</div>
+  }
+
+  return (
+    <Routes>
+      <Route index element={<New address={address} />} />
```
Comment (adamgall) [inline src/components/DAOs/DAO/Proposals/index.tsx:14]: @tbwebb22 just remove this line for now plz. Let's not render the `new` components unless `/new` is in the URL

**Review by adamgall — CHANGES_REQUESTED** [review]
Looks great @tbwebb22 , thank you. One small change then we'll merge.

**Review by adamgall — APPROVED** [review]
Noice

**Pull request merged**